### PR TITLE
Finer variables for OAuth2 redirectUriTemplate expansion

### LINKED ISSUE #6239: Add scheme/protocol variable for OAuth2 redirectUriTemplate
Currently most of us use the default redirect URL: 
```
{baseUrl}/{action}/oauth2/code/{registrationId}
```
These variables are populated in *DefaultOAuth2AuthorizationRequestResolver#expandRedirectUri*.

Due to our platform setup and constraints we need all our redirect URIs to start with **https://** 
We would like to reuse the default URL but change only the protocol part.

Suggestion: add variables **scheme**, **serverName** and **serverPort**  (naming aligned with *ServletRequest* for redirect URI templating.

### PR DESCRIPTION
**Pull request by zeratul021**
Fixes #6239

Provided baseScheme, baseHost, basePort and basePath variables that comprise the baseUrl variable.
To avoid errors during templating, I've added all variables with empty string default.

### CODE REVIEW

**Comment by zeratul021** [comment]
@rwinch I've managed to craft something with some assumptions, expecting feedback.

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolverTests.java [hunk 1]
```diff
@@ -50,6 +51,7 @@
 	public void setUp() {
 		this.registration1 = TestClientRegistrations.clientRegistration().build();
 		this.registration2 = TestClientRegistrations.clientRegistration2().build();
+		this.registration3 = TestClientRegistrations.clientRegistration3().build();
```
Comment (rwinch) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolverTests.java]: The registration1 doesn't care about the value other than it is different than registration. For this case, registration3 needs to use the new variables, so (if it is kept around) it should be named more explicitly.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java [hunk 2]
```diff
@@ -154,23 +156,52 @@ else if (AuthorizationGrantType.IMPLICIT.equals(clientRegistration.getAuthorizat
...
-		uriVariables.put("baseUrl", baseUrl);
+				.fragment(null)
+				.build();
+		String scheme = uriComponents.getScheme();
+		uriVariables.put("baseScheme", scheme == null ? "" : scheme);
+		String host = uriComponents.getHost();
+		uriVariables.put("baseHost", host == null ? "" : host);
+		// following logic is based on HierarchicalUriComponents#toUriString()
+		int port = uriComponents.getPort();
+		uriVariables.put("basePort", port == -1 ? "" : ":" + port);
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java]: Add a test for -1

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java [hunk 3]
```diff
@@ -149,20 +152,49 @@ private String resolveRegistrationId(HttpServletRequest request) {
...
+				.replaceQuery(null)
+				.fragment(null)
+				.build();
+		String scheme = uriComponents.getScheme();
+		uriVariables.put("baseScheme", scheme == null ? "" : scheme);
+		String host = uriComponents.getHost();
+		uriVariables.put("baseHost", host == null ? "" : host);
+		// following logic is based on HierarchicalUriComponents#toUriString()
+		int port = uriComponents.getPort();
+		uriVariables.put("basePort", port == -1 ? "" : ":" + port);
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: Add test for -1 port

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/registration/TestClientRegistrations.java [hunk 4]
```diff
@@ -55,6 +55,21 @@
 				.clientSecret("client-secret");
 	}
 
+	public static ClientRegistration.Builder clientRegistration3() {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/registration/TestClientRegistrations.java]: Move this to be a private static method in `DefaultOAuth2AuthorizationRequestResolverTests`

Comment (zeratul021) [reply oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/registration/TestClientRegistrations.java]: OK

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolverTests.java (hunk 1 above)
Comment (zeratul021) [reply oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolverTests.java]: OK

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java (hunk 2 above)
Comment (zeratul021) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java]: I see the test class contains much less tests then non-reactive counterpart. Also, I've already moved the client registration to that specific class as requested. Would it suffice to have the test in one place?

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java (hunk 3 above)
Comment (zeratul021) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: OK

**Comment by rwinch** [comment]
Thanks for the PR @zeratul021! This is now merged into master

**Pull request merged**